src/Week7/HW7.hs

```diff
@@ -0,0 +1,240 @@
+shuffle :: Vector a -> Rnd (Vector a)
+shuffle vt = helper (pure vt) (length vt - 1)
+  where helper v 0 = v
+        helper v n = helper (swapV' n <$> getRandomR (0, n-1) <*> v) (n - 1)
+
+-- Exercise 6 -----------------------------------------
+-- Implement a function that partitions a vector around the element at a given index
+
+-- Appending an element to a vector is O(N), so we first convert the input to a list (to use O(1) cons) and then convert the result back to a vector at the end
+partitionAt :: Ord a => Vector a -> Int -> (Vector a, a, Vector a)
+partitionAt v i = toTupleV $ helper [] (v!i) [] i (V.toList v)
+  where helper lt p gt _ []     = (lt, p, gt)
+        helper lt p gt 0 (_:xs) = helper lt p gt (-1) xs --skip the pivot
+        helper lt p gt n (x:xs) = if x < p then helper (x:lt) p gt (n-1) xs
+                                         else helper lt p (x:gt) (n-1) xs
+
+toTupleV :: ([a], a, [a]) -> (Vector a, a, Vector a)
+toTupleV (fs, i, sn) = (V.fromList fs, i, V.fromList sn)
+
+
+-- Exercise 7 -----------------------------------------
+-- Define a naive quicksort for vectors
+
+-- Quicksort (given to us)
+quicksort :: Ord a => [a] -> [a]
+quicksort [] = []
+quicksort (x:xs) = quicksort [ y | y <- xs, y < x ]
+                   <> (x : quicksort [ y | y <- xs, y >= x ])
+
+
+qsort :: Ord a => Vector a -> Vector a
+qsort v = V.fromList $ quicksort $ V.toList v
+
+-- or, using vectors directly (inefficient)
+qsort' :: Ord a => Vector a -> Vector a
+qsort' v = case V.uncons v of
+   Nothing     -> V.empty
+   Just (a,as) -> qsort' [x | x <- as, x < a]
+                  <> V.cons a (qsort' [x | x <- as, x >= a])
+-- Is there a way to use LambdaCase when we are pattern matching a function
+-- of the parameter (e.g. V.uncons v) rather than the parameter itself?
+
+-- Exercise 8 -----------------------------------------
+-- Implement randomized quicksort
+qsortR :: Ord a => Vector a -> Rnd (Vector a)
+qsortR v
+  |null v    = return V.empty
+  |otherwise = partitionRd v >>= consPartition
+
+-- helper function for further calls of quicksort, and also concatenates the sorted subvectors
+consPartition :: Ord a => (Vector a, a, Vector a) -> Rnd (Vector a)
+consPartition (v1, p, v2) = liftM2 (<>) (qsortR v1) (liftM' (V.cons p) $ qsortR v2)
+
+-- partitions the vector choosing a random element as pivot
+partitionRd :: Ord a => Vector a -> Rnd (Vector a, a, Vector a)
+partitionRd v = liftM' (partitionAt v) (getRandomR (0, length v - 1))
+
+-- Exercise 9 -----------------------------------------
+-- Implement a randomised algorithm to select the element with rank i in an unsorted array
+
+select :: Ord a => Int -> Vector a -> Rnd (Maybe a)
+select i v
+  |i >= length v || i < 0 = return Nothing
+  |otherwise = partitionRd v >>= selectSubvector i
+
+selectSubvector :: Ord a => Int -> (Vector a, a, Vector a) -> Rnd (Maybe a)
+selectSubvector i (v1, p, v2)
+  |i < l  = select i v1
+  |i == l = pure (Just p)
+  |i > l  = select (i-l-1) v2
```

Comment:
I think GHC is giving warnings here and you can fix it by replacing the last condition with otherwise.
Ignore this comment if there are no warnings.